pkg/controller/apiextensions/publication/reconciler.go

```diff
@@ -164,159 +293,172 @@ func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error)
+					log.Debug(errDeleteCR, "error", err)
+					r.record.Event(p, event.Warning(reasonDeletePub, err))
+					p.Status.SetConditions(runtimev1alpha1.ReconcileError(errors.Wrap(err, errDeleteCR)))
+					return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, p), errUpdateStatus)
 				}
 			}
-			log.Debug(waitingInstanceDeletion, "info", err)
-			p.Status.SetConditions(runtimev1alpha1.ReconcileSuccess().WithMessage(waitingInstanceDeletion))
-			return reconcile.Result{RequeueAfter: 3 * time.Second}, errors.Wrap(r.client.Status().Update(ctx, p), errUpdateInfraPubStatus)
+
+			// We requeue to confirm that all the custom resources we just
+			// deleted are actually gone. We need to requeue after a tiny wait
+			// because we won't be requeued implicitly when the CRs are deleted.
+			log.Debug(waitCRDelete)
+			r.record.Event(p, event.Normal(reasonDeletePub, waitCRDelete))
+			p.Status.SetConditions(runtimev1alpha1.ReconcileSuccess().WithMessage(waitCRDelete))
+			return reconcile.Result{RequeueAfter: tinyWait}, errors.Wrap(r.client.Status().Update(ctx, p), errUpdateStatus)
 		}
-		// Controller should be stopped before the deletion of CRD so that it
-		// doesn't crash.
-		r.ctrl.Stop(p.GetName())
+
+		// The controller should be stopped before the deletion of CRD so that
+		// it doesn't crash.
+		r.publication.Stop(requirement.ControllerName(p.GetName()))
+
 		if err := r.client.Delete(ctx, crd); resource.IgnoreNotFound(err) != nil {
 			log.Debug(errDeleteCRD, "error", err)
+			r.record.Event(p, event.Warning(reasonDeletePub, err))
 			p.Status.SetConditions(runtimev1alpha1.ReconcileError(errors.Wrap(err, errDeleteCRD)))
-			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, p), errUpdateInfraPubStatus)
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, p), errUpdateStatus)
 		}
 
+		// We should be requeued implicitly because we're watching the
+		// CustomResourceDefinition that we just deleted, but we requeue after
+		// a tiny wait just in case the CRD isn't gone after the first requeue.
+		log.Debug("Stopped requirement controller and deleted CustomResourceDefinition")
+		r.record.Event(p, event.Normal(reasonDeletePub, "Stopped requirement controller and deleted CustomResourceDefinition"))
 		p.Status.SetConditions(runtimev1alpha1.ReconcileSuccess())
-		return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, p), errUpdateInfraPubStatus)
+		return reconcile.Result{RequeueAfter: tinyWait}, errors.Wrap(r.client.Status().Update(ctx, p), errUpdateStatus)
 	}
 
-	// We will do some operations now that we will have to cleanup, so, we need
-	// a finalizer to have the chance.
-	if err := r.AddFinalizer(ctx, p); err != nil {
+	// Note that we add a finalizer to the InfrastructureDefinition that we
```

Comment:
I had a similar thought about "publication" being a field/flag of the definition rather than a distinct CR. It might be worth exploring these approaches post-release.